---
title: "Dystonia - Brain Span Analyses testing"
author: "Darren Cameron"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: true
---

## Data description

We have gene expression data and metadata for:

+ 39 individuals
+ ~~481~~ ~~438~~ 447 independent dissections (see `dystonia_brain_span_prep_data.r`)
+ RPKM correct expression values
+ ~~12~~ 7 dev stages
+ ~~8~~ ~~6~~ 8 brain regions
+ Covariates: `sex`, `rin`, `ethnicity` 

```{r setup, include=FALSE}
library(pheatmap)
library(cowplot)
library(reshape2)
library(performance) # LM diagnostic plots
library(dlookr)
library(explore)
library(car)
library(lmtest) 
library(preprocessCore) # Quantile normalisation
library(bestNormalize) # 
library(MASS) # LAD model
library(flextable)
library(tidyverse)


root_dir <- '~/Desktop/dystonia_snRNAseq_2024/'
results_dir <- paste0(root_dir, 'results/')
bulk_dir <- paste0(results_dir, '02Bulk_data/')
resources_dir <- paste0(root_dir, 'resources/sheets/')
brain_span_dir <- paste0(root_dir, 'resources/public_data/brain_span/genes_matrix_csv/')
dystonia_genes <- c("ADCY5", "ANO3", "ATP1A3", "DNAJC12", "EIF2AK2", "GCH1", "GNAL", 
                    "GNAO1", "HPCA", "KCNA1", "KCNMA1", "KCTD17", "KMT2B", "PNKD", 
                    "PRKRA", "PRRT2", "SCN8A", "SGCE", "SLC2A1", "SPR", "TH", "THAP1", 
                    "TOR1A", "TUBB4A", "VPS16")
dystonia_genes_ensembl <- c("ENSG00000173175", "ENSG00000134343", "ENSG00000105409", 
                            "ENSG00000108176", "ENSG00000055332", "ENSG00000131979", 
                            "ENSG00000141404", "ENSG00000087258", "ENSG00000121905", 
                            "ENSG00000111262", "ENSG00000156113", "ENSG00000100379", 
                            "ENSG00000167548", "ENSG00000127838", "ENSG00000180228", 
                            "ENSG00000167371", "ENSG00000196876", "ENSG00000127990", 
                            "ENSG00000117394", "ENSG00000116096", "ENSG00000180176", 
                            "ENSG00000131931", "ENSG00000136827", "ENSG00000104833", 
                            "ENSG00000215305")
brain_levels <- c("PFC", "PMSC", "NPFC", "Str", "Cer", "Hip", "Tha", "Amy")
dev_levels <- c("EarlyFetal", "MidFetal", "LateFetal", "Infancy", "Childhood",
                "Adolescence", "Adulthood")

##  Load Data  ------------------------------------------------------------------------
expr_tbl <- readRDS(paste0(bulk_dir, 'dystonia_brain_span_clean_tbl.rds')) %>%
  dplyr::select(donor, region, dev_stage, ethnicity, rin, sex, sort(names(.)[7:31]))

expr_all_tbl <- readRDS(paste0(bulk_dir, 'dystonia_brain_span_clean_all_tbl.rds')) 
```

***

## Step 1: Linear models covariate correction (yeojohn2) {.tabset}

+ Run step 1 with Yeo-Johnson transformations

```{r step 1.1, echo=FALSE, warning=FALSE, message=FALSE}
all_yeojohn_lm_lst <- list()
all_yeojohn_lm_res_lst <- list()
assumptions_yeojohn_lm_lst <- list()
gq_test_yeojohn_lst <- list()
ks_test_yeojohn_lst <- list()

# yeojohn transform - data are already yeojohn transformed
exp_yeojohn_tbl <- expr_tbl |> 
   mutate(across(ADCY5:VPS16, ~ yeojohnson(.x)$x.t))

for (gene in c(dystonia_genes)) {

  # Run linear model for dystonia genes
  linear_model <- lm(paste0(gene, '~', 'rin + ethnicity + sex'), data = exp_yeojohn_tbl)
  lm_residuals <- linear_model$residuals
  
  # Run tests to check assumptions
  model_assumptions <- check_model(linear_model)
  gq_test <- broom::tidy(lmtest::gqtest(linear_model)) |> # Goldfeld-Quant test for heteroscedasticity
    mutate(gene = gene) |>
    select(gene, statistic, p.value)
    
  ks_test <- broom::tidy(ks.test(lm_residuals, "pnorm", 
                                 mean = mean(lm_residuals), 
                                 sd = sd(lm_residuals))) |> # Kolmogorov-Smirnov Test test for normality of residuals
    mutate(gene = gene) |>
    select(gene, statistic, p.value)
      
  linear_model <- broom::tidy(linear_model)

  # Add gene summary (with name) to list
  all_yeojohn_lm_lst[[paste0(gene, '_lm_all')]] <- linear_model
  all_yeojohn_lm_res_lst[[paste0(gene, '_res')]] <- lm_residuals
  assumptions_yeojohn_lm_lst[[paste0(gene)]] <- model_assumptions
  gq_test_yeojohn_lst[[paste0(gene)]] <- gq_test
  ks_test_yeojohn_lst[[paste0(gene)]] <- ks_test
} 

```

***

### Linear model

```{r model_1.1_data, layout="l-body-outset"}
data <- bind_rows(all_yeojohn_lm_lst) |>
  mutate(across(where(is.numeric), 
                 ~ formatC(., format = "e", digits = 2))) |>
  mutate(gene = rep(str_split_i(names(all_yeojohn_lm_lst), '_', 1), each = 7)) |>
  relocate(gene)
rmarkdown::paged_table(data, options = list(rows.print = 7))
```

***

### Diagnostic plots {.tabset}

#### LM assumption plots  {.tabset}

```{r diagnostic_plots_model_1.1, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_yeojohn_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_yeojohn_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Residual normality plots {.tabset}

```{r residual_norm_plots_1.1, results='asis', echo=FALSE, warnings=FALSE}
residuals_df <- bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |>
  rename_with(~ str_replace(., "_res", ""))
  
for(i in colnames(residuals_df)){
  cat('#####',i,' \n')
  plot_normality(residuals_df, i)
  cat('\n\n\n')
}
```

***

### Diagnostic tests {.tabset}

#### Goldfeld-Quant test for heteroscedasticity

+ Genes with significant p-values: `r length(bind_rows(gq_test_yeojohn_lst) |> filter(p.value < 0.05) |> arrange(gene) |> pull(gene))` 

```{r gq_diagnostic_1.1}
data <- bind_rows(gq_test_yeojohn_lst) |>
  arrange(p.value)

gq_genes <- bind_rows(gq_test_yeojohn_lst) |>
  filter(p.value < 0.05) |>
  arrange(gene) |>
  pull(gene)

gq_genes

rmarkdown::paged_table(data, options = list(rows.print = 10))
```

***

#### Kolmogorov-Smirnov test for normality

+ Genes with significant p-values: `r length(bind_rows(ks_test_yeojohn_lst) |> filter(p.value < 0.05) |> arrange(gene) |> pull(gene))`

```{r ks_diagnostic_1.1}
data <- bind_rows(ks_test_yeojohn_lst) |>
  arrange(p.value)

bind_rows(ks_test_yeojohn_lst) |>
  filter(p.value < 0.05) |>
  arrange(gene) |>
  pull(gene)

rmarkdown::paged_table(data, options = list(rows.print = 10))
```

***

## Step 1.5: Combine residuals (all genes)

+ Prep data for step 2
+ `lm()` handles catagorical variables (`region`, `dev_stage`) automatically
+ Coerce residuals from step 1 into single expression column using `pivot_longer()`
+ Table below shows input for step 2:

***

```{r step_1.8}
expr_res_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble())

step1_tbl <- expr_res_tbl |> # Add step 1 residuals to tbl
  pivot_longer(ADCY5_res:VPS16_res, names_to = "gene", values_to = "step1_residuals") 
glimpse(step1_tbl)
```

***

## Step 2: Linear models 

***

### Model 1 {.tabset}

+ Model 1: `lm(step1_residuals ~ PFC + dev_stage + donor, data)` 

#### Plot

```{r model_1, fig.dim=c(5,3)}
regional_lm_lst <- list()
regional_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  pivot_longer(ADCY5_res:VPS16_res, names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ region - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'region', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each region
for (region in unique(step1_yeoJohn_tbl$region)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', region, '+ dev_stage + donor'), data = step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  regional_lm_lst[[paste0(region, '_lm')]] <- linear_model
  regional_lm_p[[paste0(region, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(region)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
regional_lms <- lapply(names(regional_lm_lst), function(model_name) {
  
  region <- str_split_i(model_name, '_', 1)
  
  df <- regional_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == region)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(regional_lms) |>
  mutate(region = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(region = factor(region, levels = rev(brain_levels))) |>
  select(-term) |>
  relocate(region)
  
region_plt <- plt_data |>
  ggplot(aes(x = region, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
  ylim(-35, 35) +
  Seurat::NoLegend()


print(region_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_1, results='asis', echo=FALSE}
for(i in names(regional_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(regional_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots {.tabset}

```{r Diagnostic_plots_model_1, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_1_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 8))
```

***

### Model 2 {.tabset}

+ Model 2: `lm(step1_residuals ~ EarlyFetal + region + donor, data)` 

#### Plot

```{r model_2, fig.dim=c(8,3)}
dev_stage_lm_lst <- list()
dev_stage_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  pivot_longer(ADCY5_res:VPS16_res, names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ dev_stage - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'dev_stage', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each Dev_stage
for (dev_stage in unique(step1_yeoJohn_tbl$dev_stage)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', dev_stage, '+ region + donor'), step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  dev_stage_lm_lst[[paste0(dev_stage, '_lm')]] <- linear_model
  dev_stage_lm_p[[paste0(dev_stage, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(dev_stage)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
dev_stage_lms <- lapply(names(dev_stage_lm_lst), function(model_name) {
  
  dev_stage <- str_split_i(model_name, '_', 1)
  
  df <- dev_stage_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == dev_stage)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(dev_stage_lms) |>
  mutate(dev_stage = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(dev_stage = factor(dev_stage, levels = rev(dev_levels))) |>
  select(-term) |>
  relocate(dev_stage)
  
dev_stage_plt <- plt_data |>
  ggplot(aes(x = dev_stage, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
  scale_fill_manual(values=c("#00BFC4")) +
  ylim(-42, 42) +
  Seurat::NoLegend()


print(dev_stage_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_2, results='asis', echo=FALSE}
for(i in names(dev_stage_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(dev_stage_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots  {.tabset}

```{r Diagnostic_plots_model_2, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_2_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 7))
```

***


## Step 3: Linear models (genes removed)

***

### Model 3 {.tabset}

+ Model 3: `lm(step1_residuals ~ PFC + dev_stage + donor, data)` 
+ Rerun model 1 with genes failing Goldfeld-Quant removed

#### Plot

```{r model_3, fig.dim=c(5,3)}
regional_lm_lst <- list()
regional_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  select(-contains(gq_genes)) |> # Keep only genes that passed Goldfeld-Quant
  pivot_longer(contains('_res'), names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ region - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'region', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each region
for (region in unique(step1_yeoJohn_tbl$region)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', region, '+ dev_stage + donor'), data = step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  regional_lm_lst[[paste0(region, '_lm')]] <- linear_model
  regional_lm_p[[paste0(region, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(region)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
regional_lms <- lapply(names(regional_lm_lst), function(model_name) {
  
  region <- str_split_i(model_name, '_', 1)
  
  df <- regional_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == region)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(regional_lms) |>
  mutate(region = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(region = factor(region, levels = rev(brain_levels))) |>
  select(-term) |>
  relocate(region)
  
region_plt <- plt_data |>
  ggplot(aes(x = region, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
  ylim(-35, 35) +
  Seurat::NoLegend()


print(region_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_3, results='asis', echo=FALSE}
for(i in names(regional_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(regional_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots {.tabset}

```{r Diagnostic_plots_model_3, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_3_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 8))
```

***

### Model 4 {.tabset}

+ Model 4: `lm(step1_residuals ~ EarlyFetal + region + donor, data)` 
+ Rerun model 2 with genes failing Goldfeld-Quant removed

#### Plot

```{r model_4, fig.dim=c(8,3)}
dev_stage_lm_lst <- list()
dev_stage_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  select(-contains(gq_genes)) |> # Keep only genes that passed Goldfeld-Quant
  pivot_longer(contains('_res'), names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ dev_stage - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'dev_stage', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each Dev_stage
for (dev_stage in unique(step1_yeoJohn_tbl$dev_stage)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', dev_stage, '+ region + donor'), step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  dev_stage_lm_lst[[paste0(dev_stage, '_lm')]] <- linear_model
  dev_stage_lm_p[[paste0(dev_stage, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(dev_stage)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
dev_stage_lms <- lapply(names(dev_stage_lm_lst), function(model_name) {
  
  dev_stage <- str_split_i(model_name, '_', 1)
  
  df <- dev_stage_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == dev_stage)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(dev_stage_lms) |>
  mutate(dev_stage = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(dev_stage = factor(dev_stage, levels = rev(dev_levels))) |>
  select(-term) |>
  relocate(dev_stage)
  
dev_stage_plt <- plt_data |>
  ggplot(aes(x = dev_stage, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
  ylim(-28, 28) +
  Seurat::NoLegend()


print(dev_stage_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_4, results='asis', echo=FALSE}
for(i in names(dev_stage_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(dev_stage_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots  {.tabset}

```{r Diagnostic_plots_model_4, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_4_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 7))
```

***


## Step 4: Linear models (donor removed)

### Model 5 {.tabset}

+ Model 5: `lm(step1_residuals ~ PFC + dev_stage, data)` 
+ Remove donor as a covariate

#### Plot

```{r model_5, fig.dim=c(5,3)}
regional_lm_lst <- list()
regional_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  pivot_longer(ADCY5_res:VPS16_res, names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ region - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'region', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each region
for (region in unique(step1_yeoJohn_tbl$region)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', region, '+ dev_stage'), data = step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  regional_lm_lst[[paste0(region, '_lm')]] <- linear_model
  regional_lm_p[[paste0(region, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(region)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
regional_lms <- lapply(names(regional_lm_lst), function(model_name) {
  
  region <- str_split_i(model_name, '_', 1)
  
  df <- regional_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == region)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(regional_lms) |>
  mutate(region = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(region = factor(region, levels = rev(brain_levels))) |>
  select(-term) |>
  relocate(region)
  
region_plt <- plt_data |>
  ggplot(aes(x = region, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
  ylim(-28, 28) +
  Seurat::NoLegend()


print(region_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_5, results='asis', echo=FALSE}
for(i in names(regional_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(regional_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots {.tabset}

```{r Diagnostic_plots_model_5, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_5_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 8))
```

***

### Model 6 {.tabset}

+ Model 6: `lm(step1_residuals ~ EarlyFetal + region, data)` 
+ Remove donor as a covariate

#### Plot

```{r model_6, fig.dim=c(5,3)}
dev_stage_lm_lst <- list()
dev_stage_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  pivot_longer(ADCY5_res:VPS16_res, names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ dev_stage - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'dev_stage', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each Dev_stage
for (dev_stage in unique(step1_yeoJohn_tbl$dev_stage)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', dev_stage, '+ region'), step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  dev_stage_lm_lst[[paste0(dev_stage, '_lm')]] <- linear_model
  dev_stage_lm_p[[paste0(dev_stage, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(dev_stage)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
dev_stage_lms <- lapply(names(dev_stage_lm_lst), function(model_name) {
  
  dev_stage <- str_split_i(model_name, '_', 1)
  
  df <- dev_stage_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == dev_stage)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(dev_stage_lms) |>
  mutate(dev_stage = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(dev_stage = factor(dev_stage, levels = rev(dev_levels))) |>
  select(-term) |>
  relocate(dev_stage)
  
dev_stage_plt <- plt_data |>
  ggplot(aes(x = dev_stage, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
 # scale_fill_manual(values = "#00BFC4") + # Need to add this as everything is sig.
  ylim(-125, 125) +
  Seurat::NoLegend()


print(dev_stage_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_6, results='asis', echo=FALSE}
for(i in names(dev_stage_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(dev_stage_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots  {.tabset}

```{r Diagnostic_plots_model_6, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_6_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 7))
```

***

## Step 5: Linear models (donor and genes removed)

### Model 7 {.tabset}

+ Model 7: `lm(step1_residuals ~ PFC + dev_stage, data)` 
+ Rerun model 5 with genes failing Goldfeld-Quant removed

#### Plot

```{r model_7, fig.dim=c(5,3)}
regional_lm_lst <- list()
regional_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  select(-contains(gq_genes)) |> # Keep only genes that passed Goldfeld-Quant
  pivot_longer(contains('_res'), names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ region - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'region', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each region
for (region in unique(step1_yeoJohn_tbl$region)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', region, '+ dev_stage'), data = step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  regional_lm_lst[[paste0(region, '_lm')]] <- linear_model
  regional_lm_p[[paste0(region, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(region)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
regional_lms <- lapply(names(regional_lm_lst), function(model_name) {
  
  region <- str_split_i(model_name, '_', 1)
  
  df <- regional_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == region)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(regional_lms) |>
  mutate(region = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(region = factor(region, levels = rev(brain_levels))) |>
  select(-term) |>
  relocate(region)
  
region_plt <- plt_data |>
  ggplot(aes(x = region, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
  ylim(-35, 35) +
  Seurat::NoLegend()


print(region_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_7, results='asis', echo=FALSE}
for(i in names(regional_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(regional_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots {.tabset}

```{r Diagnostic_plots_model_7, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_7_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 8))
```

***

### Model 8 {.tabset}

+ Model 8: `lm(step1_residuals ~ EarlyFetal + region + donor, data)` 
+ Rerun model 6 with genes failing Goldfeld-Quant removed

#### Plot

```{r model_8, fig.dim=c(8,3)}
dev_stage_lm_lst <- list()
dev_stage_lm_p <- list()
assumptions_lm_lst <- list()

# Create input tbl from yeoJohnson corrected residuals
step1_yeoJohn_tbl <- expr_tbl |>
  dplyr::select(dev_stage, region, donor) |>
  bind_cols(all_yeojohn_lm_res_lst |> as_tibble()) |> # Add step 1 residuals to tbl
  select(-contains(gq_genes)) |> # Keep only genes that passed Goldfeld-Quant
  pivot_longer(contains('_res'), names_to = "gene", values_to = "step1_residuals") 

# Add dummy variables to retain all dev stages
step1_yeoJohn_tbl <- model.matrix(~ dev_stage - 1, data = step1_yeoJohn_tbl) |>
  as_tibble() |>
  rename_with(~ str_replace(., 'dev_stage', '')) |>
  bind_cols(step1_yeoJohn_tbl) 

# Run model for each Dev_stage
for (dev_stage in unique(step1_yeoJohn_tbl$dev_stage)) {
  
  linear_model <- lm(paste0('step1_residuals', '~', dev_stage, '+ region'), step1_yeoJohn_tbl)
  model_assumptions <- check_model(linear_model)
  
  lm_pval <- summary(linear_model)$coefficients[2,]
  linear_model <- broom::tidy(linear_model)
  
  # Add gene summary (with name) to list
  dev_stage_lm_lst[[paste0(dev_stage, '_lm')]] <- linear_model
  dev_stage_lm_p[[paste0(dev_stage, '_P')]] <- lm_pval
  assumptions_lm_lst[[paste0(dev_stage)]] <- model_assumptions

}
  
# Pull out region line of lms and retain item name
dev_stage_lms <- lapply(names(dev_stage_lm_lst), function(model_name) {
  
  dev_stage <- str_split_i(model_name, '_', 1)
  
  df <- dev_stage_lm_lst[[model_name]]
  intercept_data <- df |> 
    filter(term == dev_stage)
  
  # Replace '(Intercept)' with the model name
  intercept_data <- intercept_data %>%
    mutate(term = model_name) 
  
  return(intercept_data)})

plt_data <- bind_rows(dev_stage_lms) |>
  mutate(dev_stage = str_replace(term, '_lm', '')) |>
  mutate(P_fdr = p.adjust(p.value, method = "fdr")) |>
  mutate(P_fdr_bool = ifelse(P_fdr < 0.05, TRUE, FALSE)) |>
  mutate(neglog10p = -log10(P_fdr)) |>
  mutate(neglog10p = ifelse(estimate < 0, -neglog10p, neglog10p)) |>
  mutate(dev_stage = factor(dev_stage, levels = rev(dev_levels))) |>
  select(-term) |>
  relocate(dev_stage)
  
dev_stage_plt <- plt_data |>
  ggplot(aes(x = dev_stage, y = neglog10p)) +
  geom_col(aes(fill = P_fdr_bool)) +
  coord_flip() +  # Flip the coordinates for better readability
  labs(x = "", y = "-log10(Padj) x sgn(B)") +
  theme_minimal() +
  geom_hline(yintercept = 0, color = "black") +
  theme(panel.grid.major.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        strip.text = element_text(size = 13),
        axis.text = element_text(size = 12)) +
  ylim(-90, 90) +
  scale_fill_manual(values=c("#00BFC4")) +
  Seurat::NoLegend()


print(dev_stage_plt)
```

***

#### LM raw {.tabset}

```{r all_LM_raw_model_8, results='asis', echo=FALSE}
for(i in names(dev_stage_lm_lst)){
  cat('#####',i,' \n')
  print(knitr::kable(dev_stage_lm_lst[[i]])) 
  cat('\n')
}
```

***

#### Diagnostic plots  {.tabset}

```{r Diagnostic_plots_model_8, results='asis', echo=FALSE, fig.dim=c(10,10)}
for(i in names(assumptions_lm_lst)){
  cat('#####',i,' \n')
  plot(assumptions_lm_lst[[i]])
  cat('\n\n')
}
```

***

#### Data

```{r model_8_data, layout="l-body-outset"}
rmarkdown::paged_table(plt_data, options = list(rows.print = 7))
```

***
